### DISCUSSION TITLE
Alternative to 401 login redirect

### DISCUSSION THREAD

**Discussion by victorarbuesmallada** [apache/superset#18929] [Q&A / Help]
**Issue**
We recently reduced the longevity of the Superset session which came with some ajax requests coming back with 401 responses.
This issue is tackled by the current [redirectUnauthorized](https://github.com/apache/superset/blob/master/superset-frontend/packages/superset-ui-core/src/connection/SupersetClientClass.ts#L166-L167) method, which basically redirects a user to the login page, which then will redirect back to the page the user was originally.

This redirection though can lead to unexpected results, since when the user is redirected back, the state is not the same.
As an example:

Sql editor when a long query is running
<img width="1431" alt="Screenshot 2022-02-24 at 11 48 20" src="https://user-images.githubusercontent.com/1241373/155519609-fda160c4-79c2-43c9-b0a8-8242952e82cc.png">

Sql editor after login redirect, current query is not shown on the main results panel and it seems that it's completed unless the user clicks on query history, in which case it's shown that it is still pending: 
![previous query](https://user-images.githubusercontent.com/1241373/155519835-2adfd3d5-d4e0-440e-98af-58c5437333df.png)
![current query](https://user-images.githubusercontent.com/1241373/155519862-74f622ab-b704-42d4-825c-11e723c5d5fc.png)

Similar issues happen when we try to save a chart, in which case we may lose some work when the redirection happens.

So pretty much the issue here is a mild annoyance when the login redirect happens, which could lead to some confusion within our users.

**Solution we'd like**
In our case, Superset has Keycloak as OIDC identity provider, which means we can have a request decorator that returns a new session when needed as part of the ajax calls responses.
Our preferred solution would be that, in order to avoid this login redirection, at the time a response is [parsed](https://github.com/geido/incubator-superset/blob/master/superset-frontend/packages/superset-ui-core/src/connection/callApi/parseResponse.ts#L33), maybe we could also refresh the cookie/session on the frontend so consequent ajax calls will be able to use a valid session.

**Comment by srinify** [comment]
Excellent feature request write up! Moving to Github Discussions so we can keep Issues focused on bugs :)

**Reply by victorarbuesmallada** [reply]
After going through this a little bit more, I guess the reason a refresh is happening is because the cookie that holds the Superset/Flask session is HttpOnly and if we were to update the cookie with an Ajax call response that would be either impossible or insecure.

I will still try to figure out if there's a way we can get around this but I don't think we will.

**Comment by victorarbuesmallada** [comment]
We actually sort this out by opening a new tab which updates the HttpOnly cookie that holds the Superset token/session. Although this is not a final solution, it actually solves the problem of losing the state when the page reloads.
Hopefully we'll come back with a better solution shortly.

**Comment by victorarbuesmallada** [comment]
Our current solution to this issue is using iframes to do the redirection.

Basically, if the user is already logged in Keycloak, an iframe hits the login endpoint, which loads a page that refreshes the superset session and then tells the parent window to close itself, which results on keeping the state.
When the Keycloak session is terminated, the user gets redirected to the login page which then redirects the user to the Keycloak login page.

This happens on a certain interval as we cannot easily modify the `SupersetClientClass.ts` file at the moment.
We'll continue with this approach unless there's a better solution.

[Here](https://gist.github.com/Painyjames/3cda47dfe3de2d646eb16a1933ac4e47) is the gist with our current changes.
